### DISCUSSION TITLE
Cortex.cpp: API Server

### DISCUSSION THREAD

**Discussion by dan-menlo** [janhq/nitro#1133] [Architecture Specs]
## Goal

- Cortex.cpp runs in the background, exposing an Open-AI compatible API server at port ? (similar to Docker Daemon)
- Other programs can connect to it as an AI provider

## Key Decisions

- Drogon or non-drogon?

## Current Architecture

- Port number?

## Related
- https://github.com/janhq/cortex.cpp/issues/1080
- #1097 
- https://github.com/janhq/cortex.cpp/issues/1010 (some timeout issue)
- https://github.com/janhq/cortex.cpp/issues/1005
- https://github.com/janhq/cortex.cpp/issues/1059
- https://github.com/janhq/cortex.cpp/issues/568
- https://github.com/janhq/cortex.cpp/issues/551
- https://github.com/janhq/cortex.cpp/issues/580

**Comment by dan-menlo** [comment]
@nguyenhoangthuan99 @vansangpfiev I recall you guys making some decisions on Drogon vs. non-Drogon a few weeks ago. Can you help me summarize:

- What was the decision about?
- Why did we eventually choose to stay with Drogon?

I can refactor them into the Spec above, once you guys provide me sufficient detail

**Reply by vansangpfiev** [reply]
> @nguyenhoangthuan99 @vansangpfiev I recall you guys making some decisions on Drogon vs. non-Drogon a few weeks ago. Can you help me summarize:
> 
> * What was the decision about?
> * Why did we eventually choose to stay with Drogon?
> 
> I can refactor them into the Spec above, once you guys provide me sufficient detail

We decided to stay with Drogon because:
- It supports 'non-blocking' socket I/O which can avoid the bottle-neck when the number of requests increase (blocking I/O fro httplib)
- It supports epoll which has better performance than select/poll (which httplib uses)
- We solved the openssl compilation issue
